### DISCUSSION TITLE
Running `promtool` with * from the docker image doesn't behave the same way with snap

### DISCUSSION THREAD

**Discussion by orfeas-k** [prometheus/prometheus#14134] [Q&A] [answered]
### What did you do?

Tried to run from the docker image a `check rules` command for every rule under a directory, both using `promtool` from `prometheus` snap package and from the image:
```bash
╰─$ promtool check rules src/prometheus_alert_rules/*.rule               
Checking src/prometheus_alert_rules/unit_unavailable.rule
  SUCCESS: 1 rules found

╰─$ docker run -v $PWD:/charm --entrypoint "/bin/promtool" prom/prometheus:latest check rules /charm/src/prometheus_alert_rules/*.rule               
zsh: no matches found: /charm/src/prometheus_alert_rules/*.rule

╰─$ docker run -v $PWD:/charm --entrypoint "/bin/promtool" prom/prometheus:latest 'check rules /charm/src/prometheus_alert_rules/*.rule'
promtool: error: expected command but got "check rules /charm/src/prometheus_alert_rules/*.rule", try --help
```

while if I run it for a specific alert rule it suceeds:
```bash
╰─$ docker run -v $PWD:/charm --entrypoint "/bin/promtool" prom/prometheus:latest check rules /charm/src/prometheus_alert_rules/unit_unavailable.rule
Checking /charm/src/prometheus_alert_rules/unit_unavailable.rule
  SUCCESS: 1 rules found
```

I get the same behaviour for `tes rules`.

### What did you expect to see?

Promtool checking all alert rules under that directory

### What did you see instead? Under which circumstances?

The following error messages (mentioned above as well)

```
╰─$ docker run -v $PWD:/charm --entrypoint "/bin/promtool" prom/prometheus:latest check rules /charm/src/prometheus_alert_rules/*.rule               
zsh: no matches found: /charm/src/prometheus_alert_rules/*.rule

╰─$ docker run -v $PWD:/charm --entrypoint "/bin/promtool" prom/prometheus:latest 'check rules /charm/src/prometheus_alert_rules/*.rule'
promtool: error: expected command but got "check rules /charm/src/prometheus_alert_rules/*.rule", try --help
```

### System information

Linux 6.5.0-35-generic x86_64

### Prometheus version

```text
promtool --version
promtool, version 2.37.0 (branch: main, revision: b41e0750abf5cc18d8233161560731de05199330)
  build user:       root@lcy02-amd64-061
  build date:       20220816-05:02:48
  go version:       go1.18.5
  platform:         linux/amd64
```

**Comment by beorn7** [comment]
I doubt this will be seen by the right audience. Maybe it's more a request for the maintainers of the snap packages?

**Comment by orfeas-k** [comment]
@beorn7 Sorry maybe it wasn't clear. Snap is behaving as I would expect it to behave while the docker image is the one "misbehaving" here.

**Comment by dgl** [comment, marked as answer]
This isn't really anything to do with the Docker image either, overriding `--entrypoint` on the Docker command line is tricky to get right, particularly as Docker supports a JSON format for multiple arguments in the Dockerfile, but not on the command line.

With:

```
docker run -v $PWD:/charm --entrypoint "/bin/promtool" prom/prometheus:latest 'check rules /charm/src/prometheus_alert_rules/*.rule'
```

The entire argument of `check rules /charm/src/prometheus_alert_rules/*.rule` is being passed as one argument, with no shell parsing it. For the globs to work a shell needs to be involved.

My suggestion to make globbing work as you expect is to map in the same path on the host as the container, i.e.:

```
docker run -v $PWD:$PWD --entrypoint "" prom/prometheus:latest promtool check rules $PWD/*.rule
Checking /home/dgl/foo.rule
  SUCCESS: 0 rules found
```

Then the globbing happens outside the container. This approach will continue working in the future as we may stop shipping a shell inside the container (i.e. use "distroless" containers).

**Comment by beorn7** [comment]
Looks like this is a usage question. I'll convert it to a discussion so that it stays around to help others.
